design/one-pager-cross-resource-referencing.md

```diff
@@ -0,0 +1,343 @@
+  n,err := GetObject(r.NetworkName)
+  if err!= nil{
+    return error
+  }
+
+  p,err := GetObject(r.ProjectName)
+  if err!= nil{
+    return error
+  }
+
+  return fmt.Sprintf("/projects/%v/global/networks/%v", n.Name, p.Name), nil
+}
+```
+
+And then update [Subnetwork type] as:
+
+```diff
+- Network string `json:"network"`
++ NetworkRef NetworkRefResolver `json:"networkRef"`
+```
+
+Finally, the YAML object for this solution can be re-written as:
+
+```yaml
+apiVersion: compute.gcp.crossplane.io/v1alpha2
+kind: Subnetwork
+metadata:
+  name: my-subnetwork
+  namespace: cool-ns
+spec:
+  ...
+  networkRef:
+    networkName: my-network
+    projectName: myproject
+  ...
+```
+
+There are a number of simplifying assumptions in this implementation. Using the
+notation we used for resources earlier:
+
+* The API group and version of `r2` and `r1` are the same
+* The namespace of `r2` and `r1` are the same
+
+#### Implementation in Crossplane
+
+To implement the above mentioned cross referencing in Crossplane, we modify the
+[Managed Reconciler] in crossplane-runtime to add `ResolveDep` stage to the
+existing stages, to have `{Observe, ResolveDep, Create, Update, Delete}` set.
+`ResolveDep` will check to see if any of the fields in the give API type are of
+interface type `ReferenceResolver`, and if so, attempts to resolve them. If
+resolution for any reason is not completed, reconciliation gets rescheduled.
+Once a reference field is resolved, its value will be stored in the equivalent
+non `Ref` field, and reconciler proceeds to the next (e.g. `Create`) stage.
+
+### Maintain High Fidelity
+
+When provisioning resources, it is desirable to support existing external
+resources which are not managed by Crossplane. For instance assume that in the
+`VPC` and `Subnet` sample solution, the `VPC` resource already exists and we
+only want to provision the `Subnet`. Since we changed the `VPCID` to `VPCIDRef`
+in the `Subnet` type, it won't be possible to use the external VPCID.
+
+To support this case, we need to keep the `VPCID` field in `Subnet`, so we
+update the modification as:
+
+```diff
+  // VPCID is the ID of the VPC.
+- VPCID string `json:"vpcId"`
++ VPCID string `json:"vpcId,omitempty"`
++ // VPCIDRef resolves the VPCID from the refenreced VPC
++ VPCIDRef VpcIDRefResolver `json:"vpcIdRef,omitempty"`
+```
+
+Note that we added `omitempty` rule to both fields, making them optional.
+
+This scheme has the following restrictions:
+
+* If none of `vpcId` or `vpcIDRef` are provided, result in an error
+* If both `vpcId` and `vpcIDRef` are provided, result in an error
+
+### Project Blocking Dependency in Resource Status
+
+Resolving a referenced attribute results in one of the following outcomes,
+ordered with higher priority:
+
+1. The referenced object is of a *different namespace* or a *different
+   group/version*. In this case the status of the resource should show an error,
+   with a message for more details.
```

Comment:
How would we determine that the referenced object existed in a different namespace or with a different group/version?

Reply:
Good point. I'll remove the namespace part.